File: certification_server/src/utils/crypto_utils.go
package utils

import (
	"crypto"
	"crypto/aes"
	"crypto/cipher"
	"crypto/ecdsa"
	"crypto/ed25519"
	"crypto/elliptic"
	"crypto/hmac"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/sha512"
	"crypto/x509"
	"encoding/base64"
	"encoding/binary"
	"encoding/pem"
	"errors"
	"fmt"
	"io"

	"golang.org/x/crypto/chacha20poly1305"
)

type KeySize int

const (
	KeySizeAES128  KeySize = 16   // 128 bits = 16 bytes
	KeySizeAES256  KeySize = 32   // 256 bits = 32 bytes
	KeySizeRSA2048 KeySize = 2048 // 2048 bits = 256 bytes
	KeySizeRSA4096 KeySize = 4096 // 4096 bits = 512 bytes
)

// 用于收纳常用的加密相关工具函数。
type CryptoUtils struct {
}

// DeriveRandomSymmetricKey 生成一个确定长度的随机的 AES 对称密钥，返回 Base64 编码的字符串形式。
func (c *CryptoUtils) DeriveRandomSymmetricKey(keySize KeySize) (string, error) {
	key, err := c.DeriveRandomSymmetricKeyBytes(keySize)
	if err != nil {
		return "", err
	}

	return base64.StdEncoding.EncodeToString(key), nil
}

// DeriveRandomSymmetricKeyBytes 生成一个确定长度的随机的 AES 对称密钥，返回原始字节切片形式。
func (c *CryptoUtils) DeriveRandomSymmetricKeyBytes(keySize KeySize) ([]byte, error) {
	if keySize != KeySizeAES128 && keySize != KeySizeAES256 {
		return nil, fmt.Errorf("unsupported symmetric key size: %d", keySize)
	}

	buf := make([]byte, int(keySize))
	if _, err := io.ReadFull(rand.Reader, buf); err != nil {
		return nil, err
	}

	return buf, nil
}

// DeriveRandomAsymmetricKey 生成一对随机的 RSA 非对称密钥。
// 返回 X.509/SPKI 格式的公钥和 PKCS#8 格式的私钥，均为 PEM 编码的字符串形式。
func (c *CryptoUtils) DeriveRandomAsymmetricKey(keySize KeySize) (string, string, error) {
	pub, pri, err := c.DeriveRandomAsymmetricKeyBytes(keySize)
	if err != nil {
		return "", "", err
	}

	return string(pub), string(pri), nil
}

// DeriveRandomAsymmetricKeyBytes 生成一对随机的 RSA 非对称密钥。
// 返回 X.509/SPKI 格式的公钥和 PKCS#8 格式的私钥，均为原始字节切片形式。
func (c *CryptoUtils) DeriveRandomAsymmetricKeyBytes(keySize KeySize) ([]byte, []byte, error) {
	if keySize != KeySizeRSA2048 && keySize != KeySizeRSA4096 {
		return nil, nil, fmt.Errorf("unsupported asymmetric key size: %d", keySize)
	}

	priv, err := rsa.GenerateKey(rand.Reader, int(keySize))
	if err != nil {
		return nil, nil, err
	}

	privDER, err := x509.MarshalPKCS8PrivateKey(priv)
	if err != nil {
		return nil, nil, err
	}
	pubDER, err := x509.MarshalPKIXPublicKey(&priv.PublicKey)
	if err != nil {
		return nil, nil, err
	}

	pubPEM := pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: pubDER})
	priPEM := pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: privDER})

	return pubPEM, priPEM, nil
}

// EncryptWithSymmetricKey 使用指定的 AES 对称密钥加密明文字符串，返回 Base64 编码的密文字符串。
func (c *CryptoUtils) EncryptWithSymmetricKey(plainText string, key []byte) (string, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	nonce := make([]byte, gcm.NonceSize())
	if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
		return "", err
	}

	cipherBytes := gcm.Seal(nil, nonce, []byte(plainText), nil)
	out := append(nonce, cipherBytes...)

	return base64.StdEncoding.EncodeToString(out), nil
}

// EncryptWithSymmetricKeyAndAAD 使用 AES-GCM 对称加密并支持附加认证数据。
func (c *CryptoUtils) EncryptWithSymmetricKeyAndAAD(plainText string, key []byte, aad []byte) (string, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	nonce := make([]byte, gcm.NonceSize())
	if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
		return "", err
	}

	cipherBytes := gcm.Seal(nil, nonce, []byte(plainText), aad)
	out := append(nonce, cipherBytes...)
	return base64.StdEncoding.EncodeToString(out), nil
}

// DecryptWithSymmetricKey 使用指定的 AES 对称密钥解密 Base64 编码的密文字符串，返回明文字符串。
func (c *CryptoUtils) DecryptWithSymmetricKey(cipherText string, key []byte) (string, error) {
	raw, err := base64.StdEncoding.DecodeString(cipherText)
	if err != nil {
		return "", err
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	if len(raw) < gcm.NonceSize() {
		return "", errors.New("invalid ciphertext length")
	}

	nonce := raw[:gcm.NonceSize()]
	cipherBytes := raw[gcm.NonceSize():]

	plain, err := gcm.Open(nil, nonce, cipherBytes, nil)
	if err != nil {
		return "", err
	}

	return string(plain), nil
}

// DecryptWithSymmetricKeyAndAAD 使用 AES-GCM 对称解密并校验附加认证数据。
func (c *CryptoUtils) DecryptWithSymmetricKeyAndAAD(cipherText string, key []byte, aad []byte) (string, error) {
	raw, err := base64.StdEncoding.DecodeString(cipherText)
	if err != nil {
		return "", err
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	if len(raw) < gcm.NonceSize() {
		return "", errors.New("invalid ciphertext length")
	}

	nonce := raw[:gcm.NonceSize()]
	cipherBytes := raw[gcm.NonceSize():]
	plain, err := gcm.Open(nil, nonce, cipherBytes, aad)
	if err != nil {
		return "", err
	}
	return string(plain), nil
}

// EncryptWithCipherSuite 按协商密码套件进行加密。
func (c *CryptoUtils) EncryptWithCipherSuite(cipherSuite string, plainText string, key []byte, aad []byte) (string, error) {
	switch cipherSuite {
	case "aes_128_gcm", "aes_256_gcm", "":
		return c.EncryptWithSymmetricKeyAndAAD(plainText, key, aad)
	case "chacha20_poly1305":
		aead, err := chacha20poly1305.NewX(key)
		if err != nil {
			return "", err
		}
		nonce := make([]byte, chacha20poly1305.NonceSizeX)
		if _, err = io.ReadFull(rand.Reader, nonce); err != nil {
			return "", err
		}
		cipherBytes := aead.Seal(nil, nonce, []byte(plainText), aad)
		out := append(nonce, cipherBytes...)
		return base64.StdEncoding.EncodeToString(out), nil
	default:
		return "", fmt.Errorf("unsupported cipher suite: %s", cipherSuite)
	}
}

// DecryptWithCipherSuite 按协商密码套件进行解密。
func (c *CryptoUtils) DecryptWithCipherSuite(cipherSuite string, cipherText string, key []byte, aad []byte) (string, error) {
	switch cipherSuite {
	case "aes_128_gcm", "aes_256_gcm", "":
		return c.DecryptWithSymmetricKeyAndAAD(cipherText, key, aad)
	case "chacha20_poly1305":
		raw, err := base64.StdEncoding.DecodeString(cipherText)
		if err != nil {
			return "", err
		}
		aead, err := chacha20poly1305.NewX(key)
		if err != nil {
			return "", err
		}
		if len(raw) < chacha20poly1305.NonceSizeX {
			return "", errors.New("invalid ciphertext length")
		}
		nonce := raw[:chacha20poly1305.NonceSizeX]
		cipherBytes := raw[chacha20poly1305.NonceSizeX:]
		plain, err := aead.Open(nil, nonce, cipherBytes, aad)
		if err != nil {
			return "", err
		}
		return string(plain), nil
	default:
		return "", fmt.Errorf("unsupported cipher suite: %s", cipherSuite)
	}
}

// EncryptWithPublicKey 使用指定的 RSA 公钥加密明文字符串，返回 Base64 编码的密文字符串。
func (c *CryptoUtils) EncryptWithPublicKey(plainText string, publicKey []byte) (string, error) {
	block, _ := pem.Decode(publicKey)
	if block == nil {
		return "", errors.New("invalid public key pem")
	}

	parsed, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return "", err
	}

	rsaPub, ok := parsed.(*rsa.PublicKey)
	if !ok {
		return "", errors.New("public key is not rsa")
	}

	enc, err := rsa.EncryptOAEP(sha256.New(), rand.Reader, rsaPub, []byte(plainText), nil)
	if err != nil {
		return "", err
	}

	return base64.StdEncoding.EncodeToString(enc), nil
}

// DecryptWithPrivateKey 使用指定的 RSA 私钥解密 Base64 编码的密文字符串，返回明文字符串。
func (c *CryptoUtils) DecryptWithPrivateKey(cipherText string, privateKey []byte) (string, error) {
	raw, err := base64.StdEncoding.DecodeString(cipherText)
	if err != nil {
		return "", err
	}

	block, _ := pem.Decode(privateKey)
	if block == nil {
		return "", errors.New("invalid private key pem")
	}

	parsed, err := x509.ParsePKCS8PrivateKey(block.Bytes)
	if err != nil {
		return "", err
	}

	rsaPriv, ok := parsed.(*rsa.PrivateKey)
	if !ok {
		return "", errors.New("private key is not rsa")
	}

	plain, err := rsa.DecryptOAEP(sha256.New(), rand.Reader, rsaPriv, raw, nil)
	if err != nil {
		return "", err
	}

	return string(plain), nil
}

// SignByAlgorithm 使用给定签名算法和私钥对消息签名，返回 Base64 签名串。
func (c *CryptoUtils) SignByAlgorithm(algorithm string, message []byte, privateKeyPEM []byte) (string, error) {
	privAny, err := parsePrivateKey(privateKeyPEM)
	if err != nil {
		return "", err
	}

	switch algorithm {
	case "ed25519":
		priv, ok := privAny.(ed25519.PrivateKey)
		if !ok {
			return "", errors.New("private key is not ed25519")
		}
		sig := ed25519.Sign(priv, message)
		return base64.StdEncoding.EncodeToString(sig), nil
	case "ecdsa_p256_sha256":
		priv, ok := privAny.(*ecdsa.PrivateKey)
		if !ok {
			return "", errors.New("private key is not ecdsa")
		}
		h := sha256.Sum256(message)
		sigDER, signErr := ecdsa.SignASN1(rand.Reader, priv, h[:])
		if signErr != nil {
			return "", signErr
		}
		return base64.StdEncoding.EncodeToString(sigDER), nil
	case "rsa_pss_sha256":
		priv, ok := privAny.(*rsa.PrivateKey)
		if !ok {
			return "", errors.New("private key is not rsa")
		}
		h := sha256.Sum256(message)
		sig, signErr := rsa.SignPSS(rand.Reader, priv, crypto.SHA256, h[:], nil)
		if signErr != nil {
			return "", signErr
		}
		return base64.StdEncoding.EncodeToString(sig), nil
	default:
		return "", fmt.Errorf("unsupported signature algorithm: %s", algorithm)
	}
}

// VerifyByAlgorithm 使用给定签名算法和公钥验证签名。
func (c *CryptoUtils) VerifyByAlgorithm(algorithm string, message []byte, signatureB64 string, publicKeyPEM []byte) error {
	sig, err := base64.StdEncoding.DecodeString(signatureB64)
	if err != nil {
		return err
	}

	pubAny, err := parsePublicKey(publicKeyPEM)
	if err != nil {
		return err
	}

	switch algorithm {
	case "ed25519":
		pub, ok := pubAny.(ed25519.PublicKey)
		if !ok {
			return errors.New("public key is not ed25519")
		}
		if !ed25519.Verify(pub, message, sig) {
			return errors.New("signature verification failed")
		}
		return nil
	case "ecdsa_p256_sha256":
		pub, ok := pubAny.(*ecdsa.PublicKey)
		if !ok {
			return errors.New("public key is not ecdsa")
		}
		h := sha256.Sum256(message)
		if !ecdsa.VerifyASN1(pub, h[:], sig) {
			return errors.New("signature verification failed")
		}
		return nil
	case "rsa_pss_sha256":
		pub, ok := pubAny.(*rsa.PublicKey)
		if !ok {
			return errors.New("public key is not rsa")
		}
		h := sha256.Sum256(message)
		return rsa.VerifyPSS(pub, crypto.SHA256, h[:], sig, nil)
	default:
		return fmt.Errorf("unsupported signature algorithm: %s", algorithm)
	}
}

// DeriveSessionKeyByHandshake 根据握手材料与协商参数派生会话密钥（Base64）。
func (c *CryptoUtils) DeriveSessionKeyByHandshake(
	keyExchange string,
	cipherSuite string,
	initiatorEphemeral string,
	responderEphemeral string,
	initiatorNonce string,
	responderNonce string,
) (string, error) {
	keyLen := 32
	if cipherSuite == "aes_128_gcm" {
		keyLen = 16
	}

	material := []byte(keyExchange + "|" + initiatorEphemeral + "|" + responderEphemeral + "|" + initiatorNonce + "|" + responderNonce)
	seed := sha512.Sum512(material)
	buf := make([]byte, keyLen)
	var counter uint32 = 1
	offset := 0
	for offset < keyLen {
		mac := hmac.New(sha256.New, seed[:])
		_ = binary.Write(mac, binary.BigEndian, counter)
		block := mac.Sum(nil)
		remain := keyLen - offset
		if remain > len(block) {
			remain = len(block)
		}
		copy(buf[offset:offset+remain], block[:remain])
		offset += remain
		counter++
	}
	return base64.StdEncoding.EncodeToString(buf), nil
}

func parsePrivateKey(privateKeyPEM []byte) (any, error) {
	block, _ := pem.Decode(privateKeyPEM)
	if block == nil {
		return nil, errors.New("invalid private key pem")
	}

	if keyAny, err := x509.ParsePKCS8PrivateKey(block.Bytes); err == nil {
		switch key := keyAny.(type) {
		case *rsa.PrivateKey, *ecdsa.PrivateKey, ed25519.PrivateKey:
			return key, nil
		}
	}

	if rsaKey, err := x509.ParsePKCS1PrivateKey(block.Bytes); err == nil {
		return rsaKey, nil
	}

	if ecKey, err := x509.ParseECPrivateKey(block.Bytes); err == nil {
		if ecKey.Curve == nil {
			ecKey.Curve = elliptic.P256()
		}
		return ecKey, nil
	}

	return nil, errors.New("unsupported private key format")
}

func parsePublicKey(publicKeyPEM []byte) (any, error) {
	block, _ := pem.Decode(publicKeyPEM)
	if block == nil {
		return nil, errors.New("invalid public key pem")
	}
	keyAny, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return nil, err
	}
	switch key := keyAny.(type) {
	case *rsa.PublicKey, *ecdsa.PublicKey, ed25519.PublicKey:
		return key, nil
	default:
		return nil, errors.New("unsupported public key type")
	}
}

File: data_server/src/utils/crypto_utils.py
import base64
import hashlib
import hmac
import os
import struct
from typing import Literal

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import (
    ec,
    ed25519,
    padding,
    rsa,
)
from cryptography.hazmat.primitives.ciphers.aead import AESGCM, ChaCha20Poly1305
from cryptography.hazmat.primitives.ciphers import aead as _aead


XChaCha20Poly1305 = getattr(_aead, "XChaCha20Poly1305", None)


KEYSIZE_AES = Literal[16, 32]  # AES-128 或 AES-256 的密钥长度（以字节为单位）
KEYSIZE_RSA = Literal[2048, 4096]  # RSA 密钥长度（以位为单位）


class CryptoUtils:
    """提供对称加密、非对称加密、签名验签等常用密码学操作的工具类。"""

    @staticmethod
    def derive_random_symmetric_key(
        key_size: KEYSIZE_AES, output_base64: bool = True
    ) -> str | bytes:
        """生成指定长度的随机对称密钥（AES）。返回 Base64 编码的字符串或原始字节数据。"""
        key = CryptoUtils.derive_random_symmetric_key_bytes(key_size)
        if output_base64:
            return base64.b64encode(key).decode("ascii")
        return key

    @staticmethod
    def derive_random_symmetric_key_bytes(key_size: KEYSIZE_AES) -> bytes:
        """生成指定长度的随机对称密钥（AES），返回原始字节。"""
        if key_size not in (16, 32):
            raise ValueError(f"unsupported symmetric key size: {key_size}")
        return os.urandom(int(key_size))

    @staticmethod
    def derive_random_asymmetric_key(
        key_size: KEYSIZE_RSA, output_base64: bool = True
    ) -> tuple[bytes, bytes] | tuple[str, str]:
        """生成指定长度的随机非对称密钥对（RSA）。
        返回 X.509 SPKI 公钥和 PKCS#8 私钥，格式为 Base64 编码字符串或原始 PEM 字节。
        """
        public_key, private_key = CryptoUtils.derive_random_asymmetric_key_bytes(
            key_size
        )
        if output_base64:
            return public_key.decode("ascii"), private_key.decode("ascii")
        return public_key, private_key

    @staticmethod
    def derive_random_asymmetric_key_bytes(
        key_size: KEYSIZE_RSA,
    ) -> tuple[bytes, bytes]:
        """生成 RSA 密钥对，返回 SPKI 公钥与 PKCS#8 私钥的 PEM 字节。"""
        if key_size not in (2048, 4096):
            raise ValueError(f"unsupported asymmetric key size: {key_size}")

        private_key = rsa.generate_private_key(
            public_exponent=65537, key_size=int(key_size)
        )
        private_pem = private_key.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.PKCS8,
            encryption_algorithm=serialization.NoEncryption(),
        )
        public_pem = private_key.public_key().public_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PublicFormat.SubjectPublicKeyInfo,
        )
        return public_pem, private_pem

    @staticmethod
    def encrypt_with_symmetric_key(plaintext: str, key: bytes) -> str:
        """使用 AES-GCM 对称密钥加密数据，返回 Base64 编码密文。"""
        return CryptoUtils.encrypt_with_symmetric_key_and_aad(plaintext, key, None)

    @staticmethod
    def encrypt_with_symmetric_key_and_aad(
        plaintext: str, key: bytes, aad: bytes | None
    ) -> str:
        """使用 AES-GCM + AAD 加密。输出为 nonce|ciphertext 的 Base64。"""
        aes_gcm = AESGCM(key)
        nonce = os.urandom(12)
        encrypted = aes_gcm.encrypt(nonce, plaintext.encode("utf-8"), aad)
        return base64.b64encode(nonce + encrypted).decode("ascii")

    @staticmethod
    def decrypt_with_symmetric_key(ciphertext: str, key: bytes) -> str:
        """使用 AES-GCM 对称密钥解密数据。"""
        return CryptoUtils.decrypt_with_symmetric_key_and_aad(ciphertext, key, None)

    @staticmethod
    def decrypt_with_symmetric_key_and_aad(
        ciphertext: str, key: bytes, aad: bytes | None
    ) -> str:
        """使用 AES-GCM + AAD 解密。"""
        raw = base64.b64decode(ciphertext)
        if len(raw) < 12:
            raise ValueError("invalid ciphertext length")
        nonce = raw[:12]
        payload = raw[12:]
        aes_gcm = AESGCM(key)
        plain = aes_gcm.decrypt(nonce, payload, aad)
        return plain.decode("utf-8")

    @staticmethod
    def encrypt_with_cipher_suite(
        cipher_suite: str, plaintext: str, key: bytes, aad: bytes | None
    ) -> str:
        """按协商密码套件加密。"""
        if cipher_suite in ("aes_128_gcm", "aes_256_gcm", ""):
            return CryptoUtils.encrypt_with_symmetric_key_and_aad(plaintext, key, aad)

        if cipher_suite == "chacha20_poly1305":
            if XChaCha20Poly1305 is not None:
                nonce = os.urandom(24)
                aead = XChaCha20Poly1305(key)
                encrypted = aead.encrypt(nonce, plaintext.encode("utf-8"), aad)
                return base64.b64encode(nonce + encrypted).decode("ascii")

            nonce = os.urandom(12)
            aead = ChaCha20Poly1305(key)
            encrypted = aead.encrypt(nonce, plaintext.encode("utf-8"), aad)
            return base64.b64encode(nonce + encrypted).decode("ascii")

        raise ValueError(f"unsupported cipher suite: {cipher_suite}")

    @staticmethod
    def decrypt_with_cipher_suite(
        cipher_suite: str, ciphertext: str, key: bytes, aad: bytes | None
    ) -> str:
        """按协商密码套件解密。"""
        if cipher_suite in ("aes_128_gcm", "aes_256_gcm", ""):
            return CryptoUtils.decrypt_with_symmetric_key_and_aad(ciphertext, key, aad)

        if cipher_suite == "chacha20_poly1305":
            raw = base64.b64decode(ciphertext)

            if XChaCha20Poly1305 is not None and len(raw) >= 24:
                nonce = raw[:24]
                payload = raw[24:]
                plain = XChaCha20Poly1305(key).decrypt(nonce, payload, aad)
                return plain.decode("utf-8")

            if len(raw) < 12:
                raise ValueError("invalid ciphertext length")
            nonce = raw[:12]
            payload = raw[12:]
            plain = ChaCha20Poly1305(key).decrypt(nonce, payload, aad)
            return plain.decode("utf-8")

        raise ValueError(f"unsupported cipher suite: {cipher_suite}")

    @staticmethod
    def encrypt_with_asymmetric_key(plaintext: str, public_key: bytes) -> str:
        """使用 RSA-OAEP(SHA256) 公钥加密。"""
        return CryptoUtils.encrypt_with_public_key(plaintext, public_key)

    @staticmethod
    def encrypt_with_public_key(plaintext: str, public_key_pem: bytes) -> str:
        """使用 RSA-OAEP(SHA256) 公钥加密。"""
        parsed = serialization.load_pem_public_key(public_key_pem)
        if not isinstance(parsed, rsa.RSAPublicKey):
            raise ValueError("public key is not rsa")

        encrypted = parsed.encrypt(
            plaintext.encode("utf-8"),
            padding.OAEP(
                mgf=padding.MGF1(algorithm=hashes.SHA256()),
                algorithm=hashes.SHA256(),
                label=None,
            ),
        )
        return base64.b64encode(encrypted).decode("ascii")

    @staticmethod
    def decrypt_with_asymmetric_key(ciphertext: str, private_key: bytes) -> str:
        """使用 RSA-OAEP(SHA256) 私钥解密。"""
        return CryptoUtils.decrypt_with_private_key(ciphertext, private_key)

    @staticmethod
    def decrypt_with_private_key(ciphertext: str, private_key_pem: bytes) -> str:
        """使用 RSA-OAEP(SHA256) 私钥解密。"""
        private_key = serialization.load_pem_private_key(private_key_pem, password=None)
        if not isinstance(private_key, rsa.RSAPrivateKey):
            raise ValueError("private key is not rsa")

        raw = base64.b64decode(ciphertext)
        plaintext = private_key.decrypt(
            raw,
            padding.OAEP(
                mgf=padding.MGF1(algorithm=hashes.SHA256()),
                algorithm=hashes.SHA256(),
                label=None,
            ),
        )
        return plaintext.decode("utf-8")

    @staticmethod
    def sign_by_algorithm(
        algorithm: str, message: bytes, private_key_pem: bytes
    ) -> str:
        """按算法签名，返回 Base64 编码签名。"""
        private_key = CryptoUtils._parse_private_key(private_key_pem)

        match algorithm:
            case "ed25519":
                if not isinstance(private_key, ed25519.Ed25519PrivateKey):
                    raise ValueError("private key is not ed25519")
                signature = private_key.sign(message)
                return base64.b64encode(signature).decode("ascii")

            case "ecdsa_p256_sha256":
                if not isinstance(private_key, ec.EllipticCurvePrivateKey):
                    raise ValueError("private key is not ecdsa")
                signature = private_key.sign(message, ec.ECDSA(hashes.SHA256()))
                return base64.b64encode(signature).decode("ascii")

            case "rsa_pss_sha256":
                if not isinstance(private_key, rsa.RSAPrivateKey):
                    raise ValueError("private key is not rsa")
                signature = private_key.sign(
                    message,
                    padding.PSS(
                        mgf=padding.MGF1(hashes.SHA256()),
                        salt_length=padding.PSS.MAX_LENGTH,
                    ),
                    hashes.SHA256(),
                )
                return base64.b64encode(signature).decode("ascii")
            case _:
                raise ValueError(f"unsupported signature algorithm: {algorithm}")

    @staticmethod
    def verify_by_algorithm(
        algorithm: str, message: bytes, signature_b64: str, public_key_pem: bytes
    ) -> None:
        """按算法验签，失败时抛出 ValueError。"""
        signature = base64.b64decode(signature_b64)
        public_key = CryptoUtils._parse_public_key(public_key_pem)

        try:
            match algorithm:
                case "ed25519":
                    if not isinstance(public_key, ed25519.Ed25519PublicKey):
                        raise ValueError("public key is not ed25519")
                    public_key.verify(signature, message)
                    return

                case "ecdsa_p256_sha256":
                    if not isinstance(public_key, ec.EllipticCurvePublicKey):
                        raise ValueError("public key is not ecdsa")
                    public_key.verify(
                        signature, message, ec.ECDSA(hashes.SHA256())
                    )
                    return

                case "rsa_pss_sha256":
                    if not isinstance(public_key, rsa.RSAPublicKey):
                        raise ValueError("public key is not rsa")
                    public_key.verify(
                        signature,
                        message,
                        padding.PSS(
                            mgf=padding.MGF1(hashes.SHA256()),
                            salt_length=padding.PSS.MAX_LENGTH,
                        ),
                        hashes.SHA256(),
                    )
                    return
        except InvalidSignature as exc:
            raise ValueError("signature verification failed") from exc

        raise ValueError(f"unsupported signature algorithm: {algorithm}")

    @staticmethod
    def derive_session_key_by_handshake(
        key_exchange: str,
        cipher_suite: str,
        initiator_ephemeral: str,
        responder_ephemeral: str,
        initiator_nonce: str,
        responder_nonce: str,
    ) -> str:
        """根据握手材料和协商参数派生会话密钥，返回 Base64。"""
        key_len = 16 if cipher_suite == "aes_128_gcm" else 32
        material = (
            f"{key_exchange}|{initiator_ephemeral}|{responder_ephemeral}|"
            f"{initiator_nonce}|{responder_nonce}"
        ).encode("utf-8")
        seed = hashlib.sha512(material).digest()

        output = bytearray()
        counter = 1
        while len(output) < key_len:
            counter_bytes = struct.pack(">I", counter)
            block = hmac.new(seed, counter_bytes, hashlib.sha256).digest()
            output.extend(block)
            counter += 1

        return base64.b64encode(bytes(output[:key_len])).decode("ascii")

    @staticmethod
    def _parse_private_key(private_key_pem: bytes):
        try:
            return serialization.load_pem_private_key(private_key_pem, password=None)
        except ValueError as exc:
            raise ValueError("unsupported private key format") from exc

    @staticmethod
    def _parse_public_key(public_key_pem: bytes):
        try:
            return serialization.load_pem_public_key(public_key_pem)
        except ValueError as exc:
            raise ValueError("invalid public key pem") from exc
